services/src/main/java/org/keycloak/services/util/DefaultClientSessionContext.java

```diff
@@ -245,16 +249,20 @@ public AuthorizationRequestContext getAuthorizationRequestContext() {
     // Loading data
 
     private boolean isAllowed(ClientScopeModel clientScope) {
-        if (isClientScopePermittedForUser(clientScope)) {
-            return true;
+        if (!isClientScopePermittedForUser(clientScope)) {
+            if (logger.isTraceEnabled()) {
+                logger.tracef("User '%s' not permitted to have client scope '%s'",
+                        clientSession.getUserSession().getUser().getUsername(), clientScope.getName());
+            }
+            return false;
         }
 
-        if (logger.isTraceEnabled()) {
-            logger.tracef("User '%s' not permitted to have client scope '%s'",
-                    clientSession.getUserSession().getUser().getUsername(), clientScope.getName());
+        if (restrictedScopes != null && !restrictedScopes.contains(clientScope.getName())) {
```

Comment:
2 nitpicks to this one:

- Maybe worth to add this check before the `isClientScopePermittedForUser` call? As `isClientScopePermittedForUser` could be relatively expensive (including DB calls for retrieve composite roles etc) and hence makes sense to rather avoid expensive call if the scope is restricted?

- The variable name `restrictedScopes` seems to be a bit confusing as those scopes looks rather those, which are not restricted. Especially interesting when the trace message below mentions `Client scope 'foo' is restricted...` just in the case when it is actually not included in the `restrictedScopes` list :-) Not sure about better name, but maybe something like `unrestrictedScopes` or `whitelistedScopes` ? This comment applies to various other method signatures and variable names...

Reply:
Done! For the second point I have maintained the `restrictedScopes` name. The idea is that those scopes are restricted or limited, no other than those are allowed. So I have changed the trace log, now it says: `Client scope 'foo' is not among the restricted scopes list and will not be processed`.